gusto/transport_terms.py

```diff
@@ -0,0 +1,205 @@
+from firedrake import dx, dot, grad, div, jump, avg, inner, \
+    outer, sign, cross, CellNormal, curl, dS
+from gusto.terms import Term
+
+
+__all__ = ["LinearAdvectionTerm", "AdvectionTerm", "VectorInvariantTerm"]
+
+
+class TransportTerm(Term):
+    """
+    Base class for transport terms.
+
+    The equation is assumed to be in the form:
+
+    q_t + L(q) = 0
+
+    where q is the (scalar or vector) field to be solved for.
+
+    :arg state: :class:`.State` object.
+    :arg test: :class:`.TestFunction` object.
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    """
+
+    def __init__(self, state, test, *, ibp="once"):
+
+        super().__init__(state, test)
+
+        self.ibp = ibp
+
+        # **** until other options are fixed ****
+        self.dS = dS
+
+
+class LinearAdvectionTerm(TransportTerm):
+    """
+    Class for linear transport equation.
+
+    :arg state: :class:`.State` object.
+    :arg test: :class:`.TestFunction` object.
+    :arg qbar: The reference function that the equation has been linearised
+               about. It is assumed that the reference velocity is zero and
+               the ubar below is the nonlinear advecting velocity
+               0.5*(u'^(n+1) + u'(n)))
+    :arg ibp: string, stands for 'integrate by parts' and can take the value
+              None, "once" or "twice". Defaults to "once".
+    :arg equation_form: (optional) string, can take the values 'continuity',
+                        which means the equation is in continuity form
+                        L(q) = div(u'*qbar), or 'advective', which means the
+                        equation is in advective form L(q) = u' dot grad(qbar).
+                        Default is "advective"
+    """
+
+    def __init__(self, state, test, qbar, ibp=None, equation_form="advective"):
+        super().__init__(state=state, test=test, ibp=ibp)
+        if equation_form == "advective" or equation_form == "continuity":
+            self.continuity = (equation_form == "continuity")
+        else:
+            raise ValueError("equation_form must be either 'advective' or 'continuity', not %s" % equation_form)
+
+        self.qbar = qbar
+
+        # currently only used with the following option combinations:
+        if self.continuity and ibp is not "once":
```

Comment:
Careful, you are lucky that python interns short strings. There is no guaranteed that `str is str.copy()`, use rather `str == str.copy()` or `str != str.copy()` as appropriate.